crates/sui-core/src/jsonrpc_index.rs

```diff
@@ -269,10 +295,28 @@ impl IndexStoreTables {
 
     #[allow(deprecated)]
     fn init(&mut self, authority_store: &AuthorityStore) -> Result<(), StorageError> {
-        // If the new coin index is empty, populate it
-        if self.coin_index_2.is_empty() {
+        let mut metadata = {
+            match self.meta.get(&()) {
+                Ok(Some(metadata)) => metadata,
+                Ok(None) | Err(_) => MetadataInfo {
+                    version: CURRENT_DB_VERSION,
+                    column_families: BTreeMap::new(),
+                },
+            }
+        };
+
+        // If the new coin index hasn't already been initialized, populate it
+        if !metadata
+            .column_families
+            .get(self.coin_index_2.cf_name())
+            .is_some_and(|cf_info| cf_info.version == CURRENT_COIN_INDEX_VERSION)
```

Comment:
should we also init it when the version is old and index is empty? it doesn't seem possible though?

Reply:
Yeah so the logic here is we re-init if either:
- the version does not match (not the !)
- the index is empty